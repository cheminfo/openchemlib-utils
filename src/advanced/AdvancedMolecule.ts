import type { Molecule } from 'openchemlib';

import { getGroupedDiastereotopicAtomIDs } from '../diastereotopic/getGroupedDiastereotopicAtomIDs.js';
import { getConnectivityMatrix } from '../util/getConnectivityMatrix.js';

import { getCanonizedDiaIDs } from './getCanonizedDiaIDs.js';
import { getDiaIDsAndH } from './getDiaIDsAndH.js';
import {
  getHeterotopicSymmetryRanks,
  getFinalRanks,
} from './getHeterotopicSymmetryRanks.js';
import { getMoleculeWithH } from './getMoleculeWithH.js';
import { getXMolecule } from './getXMolecule.js';

/**
 * This class deals with diastereotopicity information and hose codes
 * It is optimized to avoid recalculation of the same information
 */
export class AdvancedMolecule {
  private readonly originalMolecule: Molecule;
  molecule: Molecule;
  idCode: string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  private Molecule: Molecule;
  private cache: any;

  constructor(molecule: Molecule) {
    this.originalMolecule = molecule;
    this.idCode = molecule.getIDCode();
    this.molecule = this.originalMolecule.getCompactCopy();
    this.molecule.ensureHelperArrays(
      molecule.getOCL().Molecule.cHelperNeighbours,
    );
    this.Molecule = this.molecule.getOCL().Molecule;
    //@ts-expect-error TODO
    this.molecule.ensureHelperArrays(this.Molecule.cHelperNeighbours);
    this.cache = {};
  }

  toMolfile() {
    return this.molecule.toMolfile();
  }

  getMolecule() {
    return this.molecule;
  }

  fromMolecule(molecule: Molecule) {
    const idCode = molecule.getIDCode();
    if (idCode !== this.idCode) {
      // no way for optimisation
      return new AdvancedMolecule(molecule);
    }
    const advancedMolecule = new AdvancedMolecule(molecule);
    advancedMolecule.cache = {
      canonizedDiaIDs: this.cache.canonizedDiaIDs,
    };
    return advancedMolecule;
  }

  get moleculeWithH() {
    if (this.cache.moleculeWithH) return this.cache.moleculeWithH;
    this.cache.moleculeWithH = getMoleculeWithH(this.molecule);
    return this.cache.moleculeWithH;
  }

  get xMolecule() {
    if (this.cache.xMolecule) return this.cache.xMolecule;
    this.cache.xMolecule = getXMolecule(this.moleculeWithH);
    return this.cache.xMolecule;
  }

  /**
   * This is related to the current moleculeWithH. The order is NOT canonized
   */
  get diaIDs() {
    if (this.cache.diaIDs) return this.cache.diaIDs;
    const diaIDs = [];
    for (let i = 0; i < this.moleculeWithH.getAllAtoms(); i++) {
      diaIDs.push(this.canonizedDiaIDs[this.finalRanks[i]]);
    }
    this.cache.diaIDs = diaIDs;
    return diaIDs;
  }

  get canonizedDiaIDs() {
    if (this.cache.canonizedDiaIDs) return this.cache.canonizedDiaIDs;
    this.cache.canonizedDiaIDs = getCanonizedDiaIDs(this);
    return this.cache.canonizedDiaIDs;
  }

  /**
   * Returns the distance matrix for the current moleculeWithH
   */
  get distanceMatrix() {
    return getConnectivityMatrix(this.moleculeWithH, { pathLength: true });
  }

  get diaIDsAndH() {
    if (this.cache.diaIDsAndH) return this.cache.diaIDsAndH;
    this.cache.diaIDsAndH = getDiaIDsAndH(this);
    return this.cache.diaIDsAndH;
  }

  /**
   * Returns symmetryRanks for all the atoms including hydrogens. Those ranks
   * deals with topicity and is related to the current moleculeWithH.
   * In order to calculate the ranks we replace all the
   * hydrogens with a X atom.
   */
  get heterotopicSymmetryRanks() {
    if (this.cache.heterotopicSymmetryRanks) {
      return this.cache.heterotopicSymmetryRanks;
    }
    this.cache.heterotopicSymmetryRanks = getHeterotopicSymmetryRanks(
      this.xMolecule,
    );
    return [...this.cache.heterotopicSymmetryRanks];
  }

  /**
   * Returns finalRanks for all the atoms including hydrogens. Those ranks
   * deals with topicity and is related to the current moleculeWithH.
   * All the atoms have a unique identifier.j
   * In order to calculate the ranks we replace all the
   * hydrogens with a X atom.
   */
  get finalRanks() {
    if (this.cache.finalRanks) return this.cache.finalRanks;
    this.cache.finalRanks = getFinalRanks(this.xMolecule);
    return this.cache.finalRanks;
  }

  toMolfileWithH() {
    return this.moleculeWithH.toMolfile();
  }

  /**
   * Returns an array of objects containing the oclID and the corresponding hydrogens and atoms
   * for the specified atomLabel (if any)
   * This always applied to the molecule with expanded hydrogens and chirality
   * @param options
   * @returns
   */
  getGroupedDiastereotopicAtomIDs(options: GroupedDiaIDsOptions = {}) {
    return groupDiastereotopicAtomIDs(this.diaIDs, this.moleculeWithH, options);
  }
}

export interface DiaIDAndH {
  oclID: string;
  hydrogens: string[];
  heavyAtom: string | undefined;
}

export interface GroupedDiaIDsOptions {
  atomLabel?: string;
}

export interface GroupedDiaID {
  counter: number;
  atoms: number[];
  oclID: string;
  atomLabel: string;
}

export function groupDiastereotopicAtomIDs(
  diaIDs: string[],
  molecule: Molecule,
  options: GroupedDiaIDsOptions = {},
) {
  const { atomLabel } = options;
  const diaIDsObject: Record<string, GroupedDiaID> = {};
  for (let i = 0; i < diaIDs.length; i++) {
    if (!atomLabel || molecule.getAtomLabel(i) === atomLabel) {
      const diaID = diaIDs[i];
      if (!diaIDsObject[diaID]) {
        diaIDsObject[diaID] = {
          counter: 0,
          atoms: [],
          oclID: diaID,
          atomLabel: molecule.getAtomLabel(i),
        };
      }
      diaIDsObject[diaID].counter++;
      diaIDsObject[diaID].atoms.push(i);
    }
  }
  return Object.keys(diaIDsObject).map((key) => diaIDsObject[key]);
}
