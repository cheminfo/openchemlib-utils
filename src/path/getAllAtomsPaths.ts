import { Molecule } from 'openchemlib';

export interface GetAllAtomsPathsOptions {
  /**
   * Maximum path length to consider. All the paths from 0 to maxPathLength will be considered
   * @default 5
   */
  maxPathLength?: number;
}

interface AtomPath {
  path: number[];
  distance: number;
}

/**
 * We need to create an array of atoms
 * that contains an array of pathLength
 * that contains an array of object
 * @param molecule
 * @param options
 * @returns
 */
export function getAllAtomsPaths(
  molecule: Molecule,
  options: GetAllAtomsPathsOptions = {},
) {
  const { maxPathLength = 5 } = options;
  const allAtomsPaths = [];
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    const oneAtomPaths: AtomPath[][] = [];
    allAtomsPaths.push(oneAtomPaths);

    let atomPaths: AtomPath[] = [];
    atomPaths.push({ path: [i], distance: 0 });
    oneAtomPaths.push(atomPaths);

    let nextIndexes = [0];
    let nextAtoms = [i];

    for (let sphere = 1; sphere <= maxPathLength; sphere++) {
      atomPaths = [];
      oneAtomPaths.push(atomPaths);
      const currentIndexes = nextIndexes;
      const currentAtoms = nextAtoms;
      nextIndexes = [];
      nextAtoms = [];

      for (let i = 0; i < currentIndexes.length; i++) {
        const atom = currentAtoms[i];
        const index = currentIndexes[i];

        const previousPath = oneAtomPaths[sphere - 1][index].path;
        for (let conn = 0; conn < molecule.getConnAtoms(atom); conn++) {
          const connectedAtom = molecule.getConnAtom(atom, conn);
          if (previousPath.includes(connectedAtom)) continue;
          nextIndexes.push(atomPaths.length);
          nextAtoms.push(connectedAtom);
          atomPaths.push({
            path: [...previousPath, connectedAtom],
            distance: sphere,
          });
        }
      }
    }
  }
  return allAtomsPaths;
}
