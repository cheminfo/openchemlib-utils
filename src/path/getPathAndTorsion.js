/**
 * Calculates the path between 2 atoms
 * @param {import('openchemlib').Molecule} molecule
 * @param {number} from - index of the first atom
 * @param {number} to - index of the end atom
 * @param {number} maxLength - maximal length of the path
 */
export function getPathAndTorsion(molecule, from, to, maxLength) {
  let originalAtoms = []; // path before renumbering
  molecule.getPath(originalAtoms, from, to, maxLength + 1);
  let torsion;
  if (originalAtoms.length === 4) {
    torsion = molecule.calculateTorsion(originalAtoms);
  }

  return {
    atoms: originalAtoms,
    from,
    to,
    torsion,
    length: originalAtoms.length - 1,
  };
}
