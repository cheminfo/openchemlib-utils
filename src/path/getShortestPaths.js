/**
 * Get the shortest path between each pair of atoms in the molecule
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.maxLength=4]
 * @returns {Array<Array>} A matrix containing on each cell (i,j) the shortest path from atom i to atom j
 */
export function getShortestPaths(molecule, options = {}) {
  const OCL = molecule.getOCL();
  const { fromLabel = '', toLabel = '', maxLength = 3 } = options;

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  const nbAtoms = molecule.getAllAtoms();
  let allShortestPaths = new Array(nbAtoms);
  for (let i = 0; i < nbAtoms; i++) {
    allShortestPaths[i] = new Array(nbAtoms);
  }

  for (let from = 0; from < nbAtoms; from++) {
    allShortestPaths[from][from] = [from];
    for (let to = from + 1; to < nbAtoms; to++) {
      if (
        (fromAtomicNumber === 0 ||
          molecule.getAtomicNo(from) === fromAtomicNumber) &&
        (toAtomicNumber === 0 || molecule.getAtomicNo(to) === toAtomicNumber)
      ) {
        let path = [];
        molecule.getPath(path, from, to, maxLength);
        if (path.length) {
          allShortestPaths[from][to] = path.slice();
          allShortestPaths[to][from] = path.reverse();
        } else {
          allShortestPaths[from][to] = null;
          allShortestPaths[to][from] = null;
        }
      } else {
        allShortestPaths[from][to] = null;
        allShortestPaths[to][from] = null;
      }
    }
  }

  return allShortestPaths;
}
