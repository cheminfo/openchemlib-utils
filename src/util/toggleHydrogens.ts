import type { Molecule } from 'openchemlib';

import { getImplicitHydrogensCount } from './getImplicitHydrogensCount';

/**
 * Toggle presence of implicity hydrogens on/off
 * @param molecule
 * @param atomID
 */
export function toggleHydrogens(molecule: Molecule, atomID: number) {
  if (getImplicitHydrogensCount(molecule, atomID) === 0) {
    const atomsToDelete = [];
    for (let i = 0; i < molecule.getAllConnAtoms(atomID); i++) {
      const connectedAtom = molecule.getConnAtom(atomID, i);
      if (molecule.getAtomicNo(connectedAtom) === 1) {
        atomsToDelete.push(connectedAtom);
      }
    }
    molecule.deleteAtoms(atomsToDelete);
  } else {
    molecule.addImplicitHydrogens(atomID);
  }
}
