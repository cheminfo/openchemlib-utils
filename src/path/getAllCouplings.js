import { tagAtom } from '../util/tagAtom';
import { makeRacemic } from '../util/makeRacemic';

/**
 * Returns an array of all the different atom diaIDs that are connected
 * {OCL}
 * {OCL.Molecule} molecule
 * {object} [options={}]
 * {string} [options.fromLabel='H']
 * {string} [options.toLabel='H']
 * {number} [options.minLength=1]
 * {number} [options.maxLength=4]
 */

function getAllCouplings(OCL, molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;
  let paths = molecule.getAllPaths({
    fromLabel,
    toLabel,
    minLength,
    maxLength,
  });

  const minSphereSize = 0;
  const maxSphereSize = 2;

  let fragment = new OCL.Molecule(0, 0);
  for (let path of paths) {
    path.info = [];
    for (let fromTo of path.fromTo) {
      let atoms = [];
      molecule.getPath(atoms, fromTo[0], fromTo[1], path.pathLength);
      let torsion;
      if (atoms.length === 4) {
        torsion = molecule.calculateTorsion(atoms);
      }
      path.info.push({
        atoms,
        torsion,
      });

      if (!path.code) {
        path.code = [];

        let tmpMolecule = molecule.getCompactCopy();
        makeRacemic(OCL, tmpMolecule);
        tagAtom(OCL, tmpMolecule, atoms[0]);
        tagAtom(OCL, tmpMolecule, atoms[atoms.length - 1]);

        let atomMask = new Array(tmpMolecule.getAllAtoms()).fill(false);
        let atomList = [];
        let max = 0;
        let min = 0;
        for (let sphere = 0; sphere <= maxSphereSize; sphere++) {
          if (max === 0) {
            for (let atom of atoms) {
              atomMask[atom] = true;
              atomList.push(atom);
              max++;
            }
          } else {
            let newMax = max;
            for (let i = min; i < max; i++) {
              let atom = atomList[i];
              for (let j = 0; j < tmpMolecule.getConnAtoms(atom); j++) {
                let connAtom = tmpMolecule.getConnAtom(atom, j);
                atomMask[connAtom] = true;
                atomList[newMax++] = connAtom;
              }
            }
            min = max;
            max = newMax;
          }
          tmpMolecule.copyMoleculeByAtoms(fragment, atomMask, true, null);
          if (sphere >= minSphereSize) {
            path.code.push(
              fragment.getCanonizedIDCode(
                OCL.Molecule.CANONIZER_ENCODE_ATOM_CUSTOM_LABELS,
              ),
            );
          }
        }
      }
    }
  }
  return paths;
}

module.exports = getAllCouplings;
