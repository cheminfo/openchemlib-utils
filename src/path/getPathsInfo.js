import { getHoseCodesForPath } from '../hose/getHoseCodesForPath';
import { getAtomsInfo } from '../util/getAtomsInfo';
import { getConnectivityMatrix } from '../util/getConnectivityMatrix';

let fragment;

/**
 *
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.minLength=1]
 * @param {string} [opions.maxLength=4]

 */
export function getPathsInfo(molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;

  const OCL = molecule.getOCL();

  if (!fragment) {
    fragment = new OCL.Molecule(0, 0);
  }

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  // we need to find all the atoms 'fromLabel' and 'toLabel'
  let atomsInfo = getAtomsInfo(molecule);

  let pathLengthMatrix = getConnectivityMatrix(molecule, {
    pathLength: true,
  });

  for (let from = 0; from < molecule.getAllAtoms(); from++) {
    atomsInfo[from].paths = [];
    for (let to = 0; to < molecule.getAllAtoms(); to++) {
      if (from !== to) {
        if (molecule.getAtomicNo(from) === fromAtomicNumber) {
          if (molecule.getAtomicNo(to) === toAtomicNumber) {
            let pathLength = pathLengthMatrix[from][to];
            if (pathLength >= minLength && pathLength <= maxLength) {
              atomsInfo[from].paths.push(
                getHoseCodesForPath(
                  molecule,
                  from,
                  to,
                  pathLength,
                  atomsInfo[to].oclID,
                ),
              );
            }
          }
        }
      }
    }
  }

  return atomsInfo;
}
