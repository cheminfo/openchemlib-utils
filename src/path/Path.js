import { getPathsInfo } from './getPathsInfo';

export const Path = {
  getPathsInfo,
};
