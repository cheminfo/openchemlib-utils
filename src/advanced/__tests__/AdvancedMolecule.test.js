import { readFileSync } from 'fs';
import { join } from 'path';

import { Molecule } from 'openchemlib';
import { describe, it, expect } from 'vitest';

import { toggleHydrogens } from '../../util/toggleHydrogens.ts';
import { AdvancedMolecule } from '../AdvancedMolecule.ts';

describe('AdvancedMolecule', () => {
  it('ethanol', () => {
    const molecule = Molecule.fromSmiles('CCCO');
    molecule.setAtomicNo(0, 1);
    const advancedMolecule = new AdvancedMolecule(molecule);
    const diaIDs = advancedMolecule.diaIDs;
    expect(diaIDs).toHaveLength(9);
    expect(diaIDs).toStrictEqual([
      'eMHAIhFHhOtdGrj@',
      'eMHAIhFIhOtdWBj@',
      'eMHAIhFJhOtdgBj@',
      'gCaHLIeIZ`GzQ@bUP',
      'gCaHHIeIZ`GzQ@bUP',
      'gCaHLIeIZ`GzQ@bUP',
      'gCaHLIeIZ`GzQ@bUP',
      'gCaHDIeIjiJ@\x7FRHDRj@',
      'gCaHDIeIjiJ@\x7FRHDRj@',
    ]);
    const diaIDsAndH = advancedMolecule.diaIDsAndH;
    expect(diaIDsAndH).toHaveLength(9);
    expect(diaIDsAndH).toMatchSnapshot()
    const molfile = advancedMolecule.toMolfile();
    expect(getMolfileAtoms(molfile)).toStrictEqual(['O', 'C', 'C', 'H']);
    const molfileWithH = advancedMolecule.toMolfileWithH();
    expect(getMolfileAtoms(molfileWithH)).toStrictEqual([
      'O',
      'C',
      'C',
      'H',
      'H',
      'H',
      'H',
      'H',
      'H',
    ]);


    function getNewMolecule() {
      const molecule = Molecule.fromSmiles('CCCC')
      molecule.setAtomicNo(2, 8);
      molecule.setAtomicNo(3, 1);
      return Molecule.fromMolfile(molecule.toMolfile());
    }

    const newAdvancedMolecule = advancedMolecule.fromMolecule(getNewMolecule());
    expect(newAdvancedMolecule.diaIDs).toHaveLength(9);
    const groupedDiaIDs = newAdvancedMolecule.getGroupedDiastereotopicAtomIDs();
    expect(groupedDiaIDs).toHaveLength(6);
    expect(groupedDiaIDs).toMatchSnapshot()
  });

  it('ethanol toggle implicit H', () => {
    const molecule = Molecule.fromSmiles('CCO');
    const advancedMolecule = new AdvancedMolecule(molecule)
    let atoms = getAtomsAndDiaInfo(advancedMolecule)
    expect(atoms).toHaveLength(3)
    expect(atoms).toMatchSnapshot()

    toggleHydrogens(molecule, 0)

    let advancedMolecule2 = advancedMolecule.fromMolecule(molecule)
    atoms = getAtomsAndDiaInfo(advancedMolecule2)
    expect(atoms).toHaveLength(6)
    expect(atoms).toMatchSnapshot()

    toggleHydrogens(molecule, 1)
    toggleHydrogens(molecule, 0)

    advancedMolecule2 = advancedMolecule.fromMolecule(molecule)
    atoms = getAtomsAndDiaInfo(advancedMolecule2)
    expect(atoms).toHaveLength(5)
    expect(atoms).toMatchSnapshot()

    const hoses = advancedMolecule2.hoseCodes;
    expect(hoses).toHaveLength(9);
    expect(hoses).toMatchSnapshot()
    expect(hoses[3]).toStrictEqual(hoses[4]);
    expect(hoses).toMatchSnapshot()
  })


  it('2-chlorobutane', () => {
    const molecule = Molecule.fromSmiles('C[C@H](Cl)CC');
    const advancedMolecule = new AdvancedMolecule(molecule);
    toggleHydrogens(molecule, 0)
    toggleHydrogens(molecule, 1)
    toggleHydrogens(molecule, 0)
    let advancedMolecule2 = advancedMolecule.fromMolecule(molecule)
    let atoms = getAtomsAndDiaInfo(advancedMolecule2)
    expect(atoms).toHaveLength(6)
    expect(atoms).toMatchSnapshot()
  });

  it('cyclosporin', () => {
    const molfile = readFileSync(
      join(__dirname, 'data/cyclosporin.mol'),
      'utf8',
    );
    const molecule = Molecule.fromMolfile(molfile);
    const advancedMolecule = new AdvancedMolecule(molecule)
    let first = Date.now();
    expect(advancedMolecule.diaIDs).toHaveLength(196);
    expect(advancedMolecule.diaIDsAndH).toHaveLength(196);
    expect(advancedMolecule.hoseCodes).toHaveLength(196);
    first = Date.now() - first;
    const copy = advancedMolecule.fromMolecule(molecule);
    let second = Date.now();
    expect(copy.diaIDs).toHaveLength(196);
    expect(copy.diaIDsAndH).toHaveLength(196);
    expect(copy.hoseCodes).toHaveLength(196);
    second = Date.now() - second;
    expect(first).toBeGreaterThan(second * 5);
  });
});

function getMolfileAtoms(molfile) {
  const lines = molfile.split(/\r?\n/);
  const atoms = [];
  for (let line of lines) {
    if (line.match(/ [A-Z][a-z]? /)) {
      atoms.push(line.replace(/^.* ([A-Z][a-z]?) .*$/, '$1'));
    }
  }
  return atoms;
}

function getMolfileAtomMapNo(molfile) {
  const lines = molfile.split(/\r?\n/);
  const atomMapNos = [];
  for (let line of lines) {
    if (line.match(/ [A-Z][a-z]? /)) {
      atomMapNos.push(line.replace(/^.* ([0-9]+) {2}0 {2}0$/, '$1'));
    }
  }
  return atomMapNos;
}

function getAtomsAndDiaInfo(advancedMolecule) {
  const molecule = advancedMolecule.molecule;
  const atomsAndDia = []
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    const atom = {
      atomicNo: molecule.getAtomicNo(i),
      ...advancedMolecule.diaIDsAndH[i]
    }
    atomsAndDia.push(atom);
  }
  return atomsAndDia
}