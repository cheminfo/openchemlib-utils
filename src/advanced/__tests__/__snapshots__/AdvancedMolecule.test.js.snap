// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdvancedMolecule > 2-chlorobutane 1`] = `
[
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gGPDALzHRVzjbHC}H\`QJh",
    ],
    "oclID": "gJPHADIMuTe@XbhOtbIpj\`",
  },
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gGPDALjHRZzjdhC}H\`QJh",
    ],
    "oclID": "gJPHADILuTe@XdhOtbQpj\`",
  },
  {
    "atomicNo": 17,
    "heavyAtom": undefined,
    "hydrogens": [],
    "oclID": "gJPHADILuTe@X\`hOtbCpfuP",
  },
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gGPDALfHRYjjThU@_iDBIU@",
      "gGPDALfHRYjjThQ@_iDBIU@",
    ],
    "oclID": "gJPHADILuTe@XbhOtbIpj\`",
  },
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gGPDALfHRUjjfHC}H\`QJh",
    ],
    "oclID": "gJPHADILuTe@XahOtbEpj\`",
  },
  {
    "atomicNo": 1,
    "heavyAtom": "gJPHADILuTe@XdhOtbQpj\`",
    "hydrogens": [],
    "oclID": "gGPDALjHRZzjdhC}H\`QJh",
  },
]
`;

exports[`AdvancedMolecule > ethanol 1`] = `
[
  {
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHHIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFHhOtdGrj@",
  },
  {
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHLIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFIhOtdWBj@",
  },
  {
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHDIeIjiJ@RHDRj@",
    ],
    "oclID": "eMHAIhFJhOtdgBj@",
  },
  {
    "heavyAtom": "eMHAIhFIhOtdWBj@",
    "hydrogens": [],
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
  {
    "heavyAtom": "eMHAIhFHhOtdGrj@",
    "hydrogens": [],
    "oclID": "gCaHHIeIZ\`GzQ@bUP",
  },
  {
    "heavyAtom": "eMHAIhFIhOtdWBj@",
    "hydrogens": [],
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
  {
    "heavyAtom": "eMHAIhFIhOtdWBj@",
    "hydrogens": [],
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
  {
    "heavyAtom": "eMHAIhFJhOtdgBj@",
    "hydrogens": [],
    "oclID": "gCaHDIeIjiJ@RHDRj@",
  },
  {
    "heavyAtom": "eMHAIhFJhOtdgBj@",
    "hydrogens": [],
    "oclID": "gCaHDIeIjiJ@RHDRj@",
  },
]
`;

exports[`AdvancedMolecule > ethanol 2`] = `
[
  {
    "atomLabel": "C",
    "atoms": [
      0,
    ],
    "counter": 1,
    "oclID": "eMHAIhFIhOtdWBj@",
  },
  {
    "atomLabel": "C",
    "atoms": [
      1,
    ],
    "counter": 1,
    "oclID": "eMHAIhFJhOtdgBj@",
  },
  {
    "atomLabel": "O",
    "atoms": [
      2,
    ],
    "counter": 1,
    "oclID": "eMHAIhFHhOtdGrj@",
  },
  {
    "atomLabel": "H",
    "atoms": [
      3,
    ],
    "counter": 1,
    "oclID": "gCaHHIeIZ\`GzQ@bUP",
  },
  {
    "atomLabel": "H",
    "atoms": [
      4,
      5,
      6,
    ],
    "counter": 3,
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
  {
    "atomLabel": "H",
    "atoms": [
      7,
      8,
    ],
    "counter": 2,
    "oclID": "gCaHDIeIjiJ@RHDRj@",
  },
]
`;

exports[`AdvancedMolecule > ethanol toggle implicit H 1`] = `
[
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHLIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFIhOtdWBj@",
  },
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHDIeIjiJ@RHDRj@",
    ],
    "oclID": "eMHAIhFJhOtdgBj@",
  },
  {
    "atomicNo": 8,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHHIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFHhOtdGrj@",
  },
]
`;

exports[`AdvancedMolecule > ethanol toggle implicit H 2`] = `
[
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHLIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFIhOtdWBj@",
  },
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHDIeIjiJ@RHDRj@",
    ],
    "oclID": "eMHAIhFJhOtdgBj@",
  },
  {
    "atomicNo": 8,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHHIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFHhOtdGrj@",
  },
  {
    "atomicNo": 1,
    "heavyAtom": "eMHAIhFIhOtdWBj@",
    "hydrogens": [],
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
  {
    "atomicNo": 1,
    "heavyAtom": "eMHAIhFIhOtdWBj@",
    "hydrogens": [],
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
  {
    "atomicNo": 1,
    "heavyAtom": "eMHAIhFIhOtdWBj@",
    "hydrogens": [],
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
]
`;

exports[`AdvancedMolecule > ethanol toggle implicit H 3`] = `
[
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHLIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFIhOtdWBj@",
  },
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHDIeIjiJ@RHDRj@",
    ],
    "oclID": "eMHAIhFJhOtdgBj@",
  },
  {
    "atomicNo": 8,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHHIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFHhOtdGrj@",
  },
  {
    "atomicNo": 1,
    "heavyAtom": "eMHAIhFJhOtdgBj@",
    "hydrogens": [],
    "oclID": "gCaHDIeIjiJ@RHDRj@",
  },
  {
    "atomicNo": 1,
    "heavyAtom": "eMHAIhFJhOtdgBj@",
    "hydrogens": [],
    "oclID": "gCaHDIeIjiJ@RHDRj@",
  },
]
`;
